### PR TITLE
New TestBed App and multiple bug fixes/cleanups

### PR DESCRIPTION
**Pull request by mackworth**
### Summary
Creates new TestBed application to manipulate essentially all properties. Also miscellaneous bug fixes, particularly around Null data values.

### Fixes Issues
This pull request fixes the following issues:  
- #282 

### Changes
The following changes are included in this pull request:  

1. New TestBed application, intended for iPad, but runs on iPhone as well. Allows manipulation of essentially all properties to test for any problems, and allow visualization of different combinations of options.
2. Adds encoding/decoding to BEMSimpleGraph to save all properties during a restore.
3. Enabled (and fixed) more compiler type checking (e.g. NSArray <BEMCircle *> *)
3. Assorted bug fixes as well (versus existing Feature branch), especially null-data related
AverageLine: 
Color should be picked up from line if not set (default nil)

Line:
 If null data and interpolation on, then extrapolate for beginning/ending nulls.
Bezier curve should be used even when only two points.
If interpolation off, then bezier line should be interrupted by gaps (but not top/bottom).
Avoid infinity result in midPoint calculation

SimpleLineGraphView:
Support restoration during startup.
NoDataLabel color should not default from Line (which defaults to white, invisible on default background)
If null value, ensure corresponding Dot isn't left on chart
If label isn't used after initially being created, ensure it's removed from view
If neither X nor Y reference lines, set line's enableRefLines to NO (although default, might have been previously YES
If Xaxis background is defaulting to colorBottom, then also use alphaBottom to match; same for Yaxis and colorTop.
Avoid infinite loop if delegate gives a zero incrementIndex for x axis
Fix one-pixel gap between yaxis and graph
Remove spurious NSLogs

### Notes
- [ ] This pull request makes breaking changes, and if so, involves the following:  
  - [ ] Public API availability  
  - [ ] Internal functionality or behavior  
- [ ] I have run and ensured that this pull request passes all XCTests  

*Include any additional notes on your pull request*
After working on another app for a while, I'm back to the one I wanted to use this for, so I've been fixing bugs in the revised version (both mine and long-standing ones). I've merged in the changes that Sam Spencer did, but I may have misunderstood some of those; please check. 

In the process of doing this, I wanted a visualization tool that would let me check various attributes beyond what the demo program or IB does. I then got carried away and essentially implemented all the properties. (Remaining unimplemented ones are touchReportFingersRequired,  autoScaleYAxis, Dashpatterns for averageLine, XAxis, Yaxis, and Gradient choices beyond a default one.) It also would be cool if it could export to the clipboard the choices you've made into ObjC (or Swift).

The only breaking API change is that I couldn't restrain myself from fixing the multiple typos of Refrence in Line. h. As this is primarily an internal API, I think that's ok, but feel free to put it back if you differ.

I don't think I've broken any real API calls, but there are some changes:
Some of the types are now decorated (e.g. NSArray <NSNumber *> *) and for nullability
alwaysDisplayPopUpAtIndex is now an NSUInteger rather than a CGFloat
Gradients are now strongly held
Added averageValueForLineGraph to delegate
Restored modifyPopupView to be available again; not clear why removed.
Added title and label properties to AverageLine API

As last time, I apologize if I've overstepped, but I really like this library and want it to be as useful as possible

### CODE REVIEW

**Comment by mackworth** [comment]
I have no idea about the CI build error; RVM and Debian?

In running the BEMSimple tests, I noticed a couple of issues.
By adding more compiler checks, it now complains about reserved macro names. Need to add `#pragma GCC diagnostic ignored "-Wreserved-id-macro" in Test
Similarly change deployment target of test to 9.0 for test 
Finally, it fails while comparing the default UIColor of the dots. This looks like either an issue of Extended color space in iOS10, or a float v double issue.

I also note that I left a spurious ` NSLog(@"VWA");` in MasterViewController

**Comment by Boris-Em** [comment]
This looks very solid @mackworth! I'll make sure to make some time and review the PR soon. Thanks a lot for your work =)

**Comment by mackworth** [comment]
Thanks; I look forward to your comments.  

More broadly, is there a current project plan?  As I'm going to be using this in my own project, I've been making some other (what I consider) improvements, and I don't know what things you all are working on, or would be interested in receiving.  For example, I've moved to a cubic Bezier (from the quadratic) to fix the wave-like pattern on linear growth curves. I also needed a more flexible X-axis (where an X value can be any float, not just integers) to display data that might be timestamped at any time. I've implemented it for my own use, but I'd need to collaborate on the API to be able to submit it as a possible PR. Finally, as I fully "get" the code now, I'd be happy to put some time in to implement other features high on your wishlist as well. Let me know.

**Comment by mackworth** [comment]
Any thoughts?

**Comment by mackworth** [comment]
Superseded by PR #300, which also incorporates the changes mentioned above. I created a separate pull request in case the addition of the pan/zoom and variableX are determined to be inappropriate for this library.  Let me know what you think and/or if you have any problems with the other build.

**Comment by Boris-Em** [comment]
Thanks for creating a new PR. And I'm really sorry for not reviewing all of this earlier. I've been busy with life/work, but truly appreciate your work and want to get at it as soon as possible.

**Comment by mackworth** [comment]
No problem.  As a first step, try running the new Testbed, preferably on an iPad to get a feel for some of the new changes. 
I'm pretty happy with its usefulness for both learning the API and testing combinations.

**Comment by Sam-Spencer** [comment]
@mackworth @Boris-Em I have begun my review of the changes and should finish in the next day or two. There are some changes & fixes which I definitely would like to see integrated into the project and that I believe will move the it in the correct direction. That being said, I still need to test and use all the changed aspects, and (after having read through the commits & change logs) there are some things of which I am skeptical.

I will also take a look at PR #300 (thank you for separating this requests).

Thank you, @mackworth, for your extensive work on the project. Your contributions are greatly appreciated!

File: Classes/BEMSimpleLineGraphView.m [hunk 1]
```diff
@@ -1394,7 +1464,8 @@ - (CGFloat)yPositionForDotValue:(CGFloat)dotValue {
 
 #pragma mark - Deprecated Methods
 
-- (NSNumber *)calculatePointValueSum {
+
```
Comment (Boris-Em) [inline Classes/BEMSimpleLineGraphView.m]: Nit: Let's remove the extra line.

File: Sample Project/TestBed/AppDelegate.m [hunk 2]
```diff
@@ -0,0 +1,59 @@
+//
+//  AppDelegate.m
+//  TestBed
+//
+//  Created by Hugh Mackworth on 3/18/17.
+//  Copyright © 2017 Boris Emorine. All rights reserved.
+//
+
+#import "AppDelegate.h"
```
Comment (Boris-Em) [inline Sample Project/TestBed/AppDelegate.m]: Let's cleanup all of the default comments and unused methods on this file.

File: Sample Project/TestBed/MasterViewController.m [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+
+@interface MasterViewController () <MSColorSelectionViewControllerDelegate, UIPopoverPresentationControllerDelegate>
+
+@property (assign) BEMLineAnimation saveAnimationSetting;
+@property (strong, nonatomic) UIColor * saveColorSetting;
+@property (strong, nonatomic) NSString * currentColorKey;
+@property (strong, nonatomic) UIView * currentColorChip;
+@end
+
+@implementation UITextField (Numbers)
```
Comment (Boris-Em) [inline Sample Project/TestBed/MasterViewController.m]: Can we put this category on its own file?

File: Sample Project/TestBed/MasterViewController.m [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+    };
+    CGGradientRef result =  CGGradientCreateWithColorComponents(colorspace, components, locations, num_locations);
+    CGColorSpaceRelease(colorspace);
+    return result;
+}
+
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view, typically from a nib.
```
Comment (Boris-Em) [inline Sample Project/TestBed/MasterViewController.m]: Let's set the title of the view controller somewhere in viewDidLoad: self.title = @"Options".
This way the back arrow will display "Options" instead of "Master" which doesn't really make sense.

**Comment by mackworth** [comment]
There you go on the requested changes.  Still don't know what to do about Travis

While I agree that this PR doesn't break any public APIs, it's still a pretty risky one, and some of the rendering is changed (for example reference lines).
   Definitely agreed on the risky; I've built it into my app, and it's running great, but it might be good to get a couple of others to try even before putting it in "Feature".  On reference lines, assuming you mean the frame ones, they were just wrong before!  (Even now, as a result of your question on "4", i noticed that the top line was still not right.

Do we want to keep both the Sample App and the TestBed app? I feel like the TestBed app is just superior at that point. @Sam-Spencer thoughts?
   I would definitely keep both; Testbed is way too complicated to be your sample. OTOH, Sample App might want to be stripped some to make it even simpler to follow as an example.

On iPhone, it's sort of hard to go back to the graph VC from the options. Could we get a button in the navigation bar to navigate to the Graph VC?
    I did most of my work on the iPad( and it's really a far better vehicle to avoid going back and forth), but agreed and done. 

Note that if you decide not to implement #300, or only some parts of it, then I'll have to move some of the fixes in #300 back into this branch.  Conversely, if you want both, then I'll have to merge these changes into that PR.  

Looking forward to more comments...

**Comment by Boris-Em** [comment]
Thank you for being so responsive @mackworth! 
I just opened a PR against your branch that should fix Travis.
I'm alright with merging your changes on the `feature` branch pretty quickly. We will need more tests before merging it to `master`, and it will definitely require a version bump.  
Keeping both the TestBed and Sample App sounds good. Your explanation makes sense.
I really think that we'll be merging #300 as well. I'd like to finish merging this branch first though.

**Comment by mackworth** [comment]
Okay, yes, looks like that travis.yml now passes. I updated it in #300 as well, and it passes there too.

**Comment by mackworth** [comment]
Any further thoughts on this PR?

**Comment by Boris-Em** [comment]
@Sam-Spencer, did you get a chance to finish your review?

File: Classes/BEMSimpleLineGraphView.m [hunk 5]
```diff
@@ -1264,55 +1257,82 @@ - (void)handleGestureAction:(UIGestureRecognizer *)recognizer {
...
         }
-    }
-
-    // ON RELEASE
-    if (recognizer.state == UIGestureRecognizerStateEnded) {
-        if ([self.delegate respondsToSelector:@selector(lineGraph:didReleaseTouchFromGraphWithClosestIndex:)]) {
-            [self.delegate lineGraph:self didReleaseTouchFromGraphWithClosestIndex:(closestDot.tag - DotFirstTag100)];
 
+        if (self.enableTouchReport && [self.delegate respondsToSelector:@selector(lineGraph:didTouchGraphWithClosestIndex:)]) {
+            [self.delegate lineGraph:self didTouchGraphWithClosestIndex:index];
```
Comment (Sam-Spencer) [inline Classes/BEMSimpleLineGraphView.m]: Not sure if this was introduced in this PR but it looks like when the data source provides a BEMNullGraphValue for a data point it also gets returned by the touch report and displayed in popup views. Something to look into... should be an easy fix.

File: Classes/BEMAverageLine.h [hunk 6]
```diff
@@ -35,7 +35,7 @@
 
 
 /// Dash pattern for the average line
-@property (strong, nonatomic, nullable) NSArray *dashPattern;
+@property (strong, nonatomic, nullable) NSArray <NSNumber *> *dashPattern;
```
Comment (Sam-Spencer) [inline Classes/BEMAverageLine.h]: Yes! Love these changes! Perfect!

File: Classes/BEMAverageLine.h [hunk 7]
```diff
@@ -11,7 +11,7 @@
 
 
 /// A line displayed horizontally across the graph at the average y-value
-@interface BEMAverageLine : NSObject
+@interface BEMAverageLine : NSObject <NSCoding>
```
Comment (Sam-Spencer) [inline Classes/BEMAverageLine.h]: Great! Love that BEMAverageLine is now also NSCoding compliant.

File: Sample Project/TestBed/ARFontPickerViewController.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+//  POSSIBILITY OF SUCH DAMAGE.
+
+@import UIKit;
+
+@class ARFontPickerViewController;
```
Comment (Sam-Spencer) [inline Sample Project/TestBed/ARFontPickerViewController.h]: I am very wary of using third-party dependencies for a demo / test app included with the project unless it is absolutely necessary. While this is a robust solution, it may be better to simply include a toggle between two hard-coded system-installed fonts.

File: Classes/BEMLine.m [hunk 9]
```diff
@@ -142,26 +143,62 @@ - (void)drawRect:(CGRect)rect {
 
     self.points = [NSMutableArray arrayWithCapacity:self.arrayOfPoints.count];
```
Comment (Sam-Spencer) [inline Classes/BEMLine.m]: Absolutely agreed. Maybe this is something we can work on after merging...

File: Sample Project/SimpleLineChartTests/CustomizationTests.m [hunk 10]
```diff
@@ -12,7 +12,7 @@
 #pragma clang diagnostic ignored "-Wfloat-equal"
 
 @interface BEMSimpleLineGraphView ()
-// Allow tester to get to internal properties
+//Allow tester to get to internal properties
```
Comment (Sam-Spencer) [inline Sample Project/SimpleLineChartTests/CustomizationTests.m]: Nit: need space between comment marker and first word.

File: Sample Project/TestBed/MSColorPicker/MSColorComponentView.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <UIKit/UIKit.h>
+
+@class MSSliderView;
```
Comment (Sam-Spencer) [inline Sample Project/TestBed/MSColorPicker/MSColorComponentView.h]: Again, I am concerned about including third-party components here when not absolutely necessary. For example, this color choice problem can be easily solved with a few lines of our own code that provide a few standard values... rather than relying on external code which may cause issues, need to be rewritten, updated, etc. at some point.

File: Sample Project/TestBed/MSColorPicker/MSColorWheelView.m [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSColorWheelView.h"
+#import "MSColorUtils.h"
+
+@interface MSColorWheelView ()
```
Comment (Sam-Spencer) [inline Sample Project/TestBed/MSColorPicker/MSColorWheelView.m]: See other comments about third-party library inclusion...

**Comment by Boris-Em** [comment]
Thank you @Sam-Spencer!

1. Agreed.  
2. Unless I misunderstood something, it doesn't seem to me that BEMSimpleLineGraph is taking on any third party dependencies: only the TestBed app is. A developer using our library wouldn't take on any additional dependency by using BEMSimpleLineGraph. I'm fine with this, as long as the external dependencies don't make it to the actual library.  
3. I agree with you that ideally using tests would be a better way to unsure nothing is getting broken. But this is something that we've been really bad at, and furthermore, testing UI components is far from easy... This is why I can see value in the TestBed app.

File: Classes/BEMSimpleLineGraphView.h [hunk 13]
```diff
@@ -516,7 +520,7 @@ IB_DESIGNABLE @interface BEMSimpleLineGraphView : UIView <UIGestureRecognizerDel
  @param graph The graph object requesting the padding value.
  @param popupView The popup view owned by the graph that needs to be modified
  @param index The index of the element associated with the popup view */
-- (void)lineGraph:(BEMSimpleLineGraphView *)graph modifyPopupView:(UIView *)popupView forIndex:(NSUInteger)index __unavailable;
+- (void)lineGraph:(BEMSimpleLineGraphView *)graph modifyPopupView:(UIView *)popupView forIndex:(NSUInteger)index;
```
Comment (mackworth) [inline Classes/BEMSimpleLineGraphView.h]: Hmm, never saw it as unavailable.  Why was it taken out? How are you supposed to use the custom Popup if it stays the same on every point?

File: Classes/BEMSimpleLineGraphView.m (hunk 5 above)
Comment (mackworth) [inline Classes/BEMSimpleLineGraphView.m]: Hmm, I don't think so; circleDots aren't created when there's no value, so closestDotfromTouchInputLine can't see them.  I tried it in TestBed and it didn't show null values.

**Comment by mackworth** [comment]
So, thanks for the nice compliments. To respond to the general points, 

On the third-party libraries, Boris is correct, there's no dependency for the library itself, just internal to TestBed. 
So on TestBed, the main uses for me are:
(a) to test all kinds of different combinations of features. Ticking them on/off and seeing how they interact is much faster than coding and recompiling, 
(b) lets one learn what all the different options mean, better than the comments do, cause you can see them work. 
(c) Finally, lets one visually create the combination desired to get the ideal graph look beyond what IB can do.  This is particularly where the third party libraries that let me choose the exact font and color I would use in my program are nice.  (this is also where it would be cool to be able to export the exact code itself).

The sample app is to see how to incorporate SimpleLineGraphView into your own program; Testbed is to see what all the possibilities are. 

But as I've said, it's your library :), so let me know.    

To summarize, I think the open questions are:

1. Including TestBed
2. Using third party (open source and included) libraries in TestBed
3. "unavailable" on modifyPopupView
4. Sam, let me know if you're really seeing BEMNulls in touchReport.  I remember I did break that once, but I don't think this version has that problem.
5. Missing space in comment.

**Comment by mackworth** [comment]
So, @Sam-Spencer, are we good on 1 and 2 above, and any further info on 3 and 4?

**Comment by Sam-Spencer** [comment]
@mackworth It looks like when a `BEMNullGraphValue` is added to the graph, those values are not handled properly in the interface. For example, if the delegate does not implement a custom popup but requests that popups be shown, the graph will print the literal value of `BEMNullGraphValue` in the popup (which happens to be `CGFLOAT_MAX`).

Interestingly enough, it seems that the graph does not display a dot when it comes across one of these values. Yet it is still also reported back to the delegate in other methods (i.e. `didTouchGraphWithClosestIndex:`).

Here's what it looks like when this happens:

simulator screen shot jul 18 2017 2 35 10 pm

I should also note that this behavior only seems to occur when doing subsequent (non-initial) graph updates (e.g. a point has been added or removed and the graph object is reloaded).

**Comment by mackworth** [comment]
Good catch!  There was a confusion whether to have the same number of circleDots as dataPoints (including Nulls) or only have one for each valid dataPoint. 
I've added a commit where they are always the same length, but uses lack of superview to distinguish the ones representing Nulls.

**Comment by Sam-Spencer** [comment]
The PR now looks good to merge into the feature branch. Thank you so much for all the work you've done here @mackworth!
